File: libtorrent_impl/torrent_state.py
from clients import TorrentState, FieldInfo
from libtorrent_impl.params import STATUS_MAPPING
from libtorrent_impl.utils import LibtorrentClientException, format_tracker_error
from models import LibtorrentTorrent


def _convert_status(state):
    return STATUS_MAPPING[int(state)]


class LibtorrentTorrentState(TorrentState):
    TRACKER_PENDING = 0
    TRACKER_ANNOUNCING = 1
    TRACKER_SUCCESS = 2
    TRACKER_ERROR = 3

    _FIELD_MAPPING = [
        FieldInfo('name', 'name'),
        FieldInfo('status', 'state', converter=_convert_status),
        FieldInfo('download_path', 'save_path'),
        FieldInfo('size', 'total_wanted'),
        FieldInfo('downloaded', 'all_time_download'),
        FieldInfo('uploaded', 'all_time_upload'),
        # Work around weird issue where libtorrent returns 1 byte/s for a long time
        FieldInfo('download_rate', 'download_payload_rate', converter=lambda i: i if i > 1 else 0),
        FieldInfo('upload_rate', 'upload_payload_rate', converter=lambda i: i if i > 1 else 0),
        FieldInfo('progress', 'progress'),
        FieldInfo('date_added', None, converter=lambda _: None),
        FieldInfo('error', 'error', converter=lambda i: i or None),
        FieldInfo('state', 'state', converter=str, public=False),
    ]

    def __init__(self, manager, handle, *, torrent_file=None, download_path=None, db_torrent=None):
        status = handle.status()

        super().__init__(manager, str(status.info_hash))

        self.handle = handle
        self.state = None
        self.tracker_status = self.TRACKER_PENDING

        if db_torrent:
            self.db_torrent = db_torrent
        else:
            self.db_torrent = self._load_or_create_db_torrent(torrent_file, download_path)

        self.update_from_status(handle.status())

    def _load_or_create_db_torrent(self, torrent_file, download_path):
        try:
            return LibtorrentTorrent.select().where(
                LibtorrentTorrent.libtorrent == self.manager.instance_config,
                LibtorrentTorrent.info_hash == self.info_hash,
            ).get()
        except LibtorrentTorrent.DoesNotExist:
            if not torrent_file or not download_path:
                raise LibtorrentClientException(
                    'Creating a new LibtorrentTorrent without supplying torrent_file and download_path.')
            return LibtorrentTorrent.create(
                libtorrent=self.manager.instance_config,
                info_hash=self.info_hash,
                torrent_file=torrent_file,
                download_path=download_path,
                resume_data=None,  # To be filled in later from the main thread
            )

    def delete(self):
        self.db_torrent.delete_instance()

    def update_from_status(self, status):
        if __debug__:
            if self.info_hash != str(status.info_hash):
                raise LibtorrentClientException('Updating wrong TorrentStatus')
        return self._sync_fields(status)

    def update_tracker_success(self):
        self.tracker_status = self.TRACKER_SUCCESS

        if self.tracker_error:
            self.tracker_error = None
            return True
        return False

    def update_tracker_error(self, alert):
        self.tracker_status = self.TRACKER_ERROR

        tracker_error = format_tracker_error(alert)
        if self.tracker_error != tracker_error:
            self.tracker_error = tracker_error
            return True
        return False

    def to_dict(self):
        result = super().to_dict()
        result['tracker_error'] = self.tracker_error
        return result

File: libtorrent_impl/speedups/TorrentState.cpp
#include <string>

#include <libtorrent/torrent_status.hpp>
#include <sqlite3.h>

#include "yuarel.hpp"
#include "SqliteHelper.hpp"
#include "TorrentState.hpp"

namespace lt = libtorrent;

Logger TorrentState::logger("TorrentState");

#define UPDATE_STATE(A, B) { auto __temp = B; if (A != __temp) { A = __temp; updated = true; } }

TorrentState::TorrentState(int64_t row_id, lt::torrent_status *status)
        : handle(status->handle)
        , row_id(row_id)
        , tracker_status(TRACKER_STATUS_PENDING)
        , status(STATUS_CHECK_WAITING)
        , downloaded(0)
        , uploaded(0)
        , download_rate(0)
        , upload_rate(0)
        , progress(0)
        , error("")
        , tracker_error("")
        , date_added(0) {
    this->info_hash = status->info_hash.to_string();
    this->name = status->name;
    this->download_path = status->save_path;
    this->size = status->total_wanted;
    this->update_from_status(status);
}

void TorrentState::insert_db_row(sqlite3 *db, std::string torrent_file, std::string download_path,
                                 std::string *name_ptr) {
    SqliteStatement stmt = SqliteStatement(
            db, "INSERT INTO torrent (info_hash, torrent_file, download_path, name) VALUES (?001, ?002, ?003, ?004)");

    std::string name;
    std::string info_hash_str = lt::to_hex(this->info_hash);

    stmt.bind_blob(1, info_hash_str.c_str());
    stmt.bind_blob(2, torrent_file);
    stmt.bind_text(3, download_path);
    if (name_ptr) {
        name = *name_ptr;
        stmt.bind_text(4, name.c_str());
    }
    this->row_id = stmt.insert();
}

void TorrentState::delete_db_row(sqlite3 *db) {
    if (this->row_id == -1) {
        TorrentState::logger.error("Trying to delete_db_row for TorrentState with id -1.");
        return;
    }
    SqliteStatement stmt = SqliteStatement(db, "DELETE FROM torrent WHERE id = ?001");
    stmt.bind_int64(1, this->row_id);
    stmt.exec();
    this->row_id = -1;
}

bool TorrentState::update_from_status(lt::torrent_status *status) {
    bool updated = false;
    this->state = status->state;
    UPDATE_STATE(this->status, get_alcazar_status(status->state));
    // Status must be stopped since it is paused
    // Alternative 'cleaner' solution would be to add a paused attribute and update that
    if (status->paused) {
        UPDATE_STATE(this->status, STATUS_STOPPED);
    }
    UPDATE_STATE(this->size, status->total_wanted);
    UPDATE_STATE(this->downloaded, status->all_time_download);
    UPDATE_STATE(this->uploaded, status->all_time_upload);
    // Hack rates, as libtorrent seems to return very small values for a while after all transfer is complete
    UPDATE_STATE(this->download_rate, status->download_payload_rate <= 2 ? 0 : status->download_payload_rate);
    UPDATE_STATE(this->upload_rate, status->upload_payload_rate <= 2 ? 0 : status->upload_payload_rate);
    UPDATE_STATE(this->progress, status->progress);
    UPDATE_STATE(this->error, status->error);
    UPDATE_STATE(this->date_added, status->added_time);
    return updated;
}

bool TorrentState::update_file_name(std::string name) {
    bool updated = false;
    UPDATE_STATE(this->name, name);
    return updated;
}

bool TorrentState::update_download_path(std::string download_path) {
    bool updated = false;
    UPDATE_STATE(this->download_path, download_path);
    return updated;
}

bool TorrentState::update_tracker_announce() {
    bool updated = false;
    UPDATE_STATE(this->tracker_status, TRACKER_STATUS_ANNOUNCING);
    return updated;
}

bool TorrentState::update_tracker_reply() {
    bool updated = false;
    UPDATE_STATE(this->tracker_status, TRACKER_STATUS_SUCCESS);
    UPDATE_STATE(this->tracker_error, "");
    return updated;
}

bool TorrentState::update_tracker_error(lt::tracker_error_alert *alert) {
    bool updated = false;
    // When pausing, there is a connection lost alert
    // This should be suppressed so the client shows paused instead of stopped with error
    if (this->handle.status().paused) {
        return false;   
    }
    UPDATE_STATE(this->tracker_status, TRACKER_STATUS_ERROR);
    UPDATE_STATE(this->tracker_error, format_tracker_error(alert));
    return updated;
}

std::unordered_map <std::string, std::string> host_from_url_cache;

std::string extract_host_from_url(std::string url) {
    auto cache_item = host_from_url_cache.find(url);
    if (cache_item != host_from_url_cache.end()) {
        return cache_item->second;
    }

    char buffer[url.size() + 1];
    url.copy(buffer, url.size());
    buffer[url.size()] = 0;

    yuarel y_url;
    std::string result;
    if (yuarel_parse(&y_url, buffer) == -1) {
        result = url;
    }
    if (y_url.host) {
        result = std::string(y_url.host);
    } else {
        result = url;
    }

    return host_from_url_cache[url] = result;
}

std::string format_tracker_error(lt::tracker_error_alert *alert) {

    std::string host = extract_host_from_url(std::string(alert->tracker_url()));
    std::string errormessage = alert->error.message();
    // Ugliest UTF-8 safe first letter uppercase code
    if (errormessage[0] >= 32 && errormessage[0] < 128) {
        errormessage[0] = std::toupper(errormessage[0]);
    }

    char buffer[400];

    snprintf(
            buffer,
            sizeof(buffer) / sizeof(buffer[0]),
            "%s - %s (%s)",
            errormessage.c_str(),
            alert->error_message(),
            host.c_str()
    );

    return std::string(buffer);
}
